src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,227 @@
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted = true;
+        return mWriter->Complete(error);
+    }
+
+    /// Complete the given command.
+    ///
+    /// Automatically handles retries for sending.
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    CHIP_ERROR Complete(CHIP_ERROR error)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = mWriter->Complete(error);
+
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingResponses());
+        return mWriter->Complete(error);
+    }
+
+    /// Sends the specified data structure as a response
+    ///
+    /// This version of the send has built-in RETRY and handles
+    /// Flush/Complete automatically.
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    template <typename ReplyData>
+    CHIP_ERROR Send(const ReplyData & data)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = data.Encode(ResponseEncoder(ReplyData::GetCommandId()));
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            LogErrorOnFailure(err);
+            err = mWriter->Complete(err);
+        }
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingResponses());
+        err = data.Encode(ResponseEncoder(ReplyData::GetCommandId()));
+
+        // If encoding fails, we will end up sending an error back to the other side
+        // the caller
+        LogErrorOnFailure(err);
+        if (err == CHIP_NO_ERROR)
+        {
+            err = mWriter->Complete(err);
+        }
+        else
+        {
+            // Error in "complete" is not something we can really forward anymore since
+            // we already got an error in Encode ... just log this.
+            LogErrorOnFailure(mWriter->Complete(err));
+        }
+
+        return err;
+    }
+
+private:
+    InvokeResponder * mWriter;
+    bool mCompleted = false;
+};
+
+enum ReplyAsyncFlags
+{
+    // Some commmands that are expensive to process (e.g. crypto)
+    // should ack right away to improve reliability and reduce
+    // needless retries.
+    kFlushAckRightAway = 0x0001,
+};
+
+class InvokeReply
+{
+public:
+    virtual ~InvokeReply() = default;
+
+    // reply with no data
+    CHIP_ERROR Reply(CHIP_ERROR e) { return this->Reply().Complete(e); }
+
+    // Send a reply "NOW" to the given invoke
```

Comment:
nit, this isn't exactly sending the reply right? This is adding the response to the reply. Sending is completely separate to this

Reply:
Updated comment.